File: Source/code/forms/formImportBitmap.pas
// ===================================================================
//
// (c) Paul Alan Freshney 2012-2022
// www.freshney.org :: paul@freshney.org :: maximumoctopus.com
//
// https://github.com/MaximumOctopus/LEDMatrixStudio
//
// Please do not modifiy this comment section

//
// ===================================================================

unit formImportBitmap;


interface


uses
  Windows, Messages, SysUtils, Variants, Classes, Graphics, Controls, Forms,
  Dialogs, ExtDlgs, ExtCtrls, StdCtrls, Buttons, Vcl.ComCtrls,

  languagehandler;


type
  TfrmImportBitmap = class(TForm)
    opdMain: TOpenPictureDialog;
    bOK: TBitBtn;
    bCancel: TBitBtn;
    Bevel1: TBevel;
    lHelpText: TLabel;
    gbSettings: TGroupBox;
    Label2: TLabel;
    Label3: TLabel;
    Label4: TLabel;
    cbWidth: TComboBox;
    cbHeight: TComboBox;
    eFrames: TEdit;
    lWidth: TLabel;
    lHeight: TLabel;
    lImageHeight: TLabel;
    lImageWidth: TLabel;
    Bevel2: TBevel;
    bAuto: TBitBtn;
    cbRGBImport: TCheckBox;
    cbCreateNew: TCheckBox;
    pcImportMethod: TPageControl;
    tsSingleImage: TTabSheet;
    tsMultipleImages: TTabSheet;
    Shape1: TShape;
    iImport: TImage;
    lFileName: TLabel;
    bSelect: TBitBtn;
    iMultipleImages: TImage;
    eMIFirstImage: TEdit;
    eMIPattern: TEdit;
    Label5: TLabel;
    Label7: TLabel;
    sbMISelectFirstImage: TSpeedButton;
    Label8: TLabel;
    eMIFirstFrame: TEdit;
    Label9: TLabel;
    Label10: TLabel;
    eMIPadLength: TEdit;
    lImageLengthExample: TLabel;
    procedure bSelectClick(Sender: TObject);
    procedure FormCreate(Sender: TObject);
    procedure bOKClick(Sender: TObject);
    procedure bCancelClick(Sender: TObject);
    procedure bAutoClick(Sender: TObject);
    procedure cbRGBImportClick(Sender: TObject);
    procedure sbMISelectFirstImageClick(Sender: TObject);
  private
    procedure SetGUILanguageText;
    procedure SetMultipleImageDetails;
  public
    ImportMode    : integer;

    RGBImport     : boolean;
    CreateNew     : boolean;
    ImageFilename : string;
    FrameCount    : integer;
    FrameWidth    : integer;
    FrameHeight   : integer;

    FirstFrame    : integer;
    PadLength     : integer;
    Pattern       : string;
  end;


const
  ImportModeInvalid        = 0;
  ImportModeSingleImage    = 1;
  ImportModeMultipleImages = 2;


var
  frmImportBitmap: TfrmImportBitmap;


implementation


{$R *.dfm}


procedure TfrmImportBitmap.FormCreate(Sender: TObject);
var
  x : integer;

begin
  ImageFilename := '';
  FrameCount    := -1;

  for x := 1 to 64 do begin
    cbWidth.Items.Add(IntToStr(x));
    cbHeight.Items.Add(IntToStr(x));
  end;

  cbWidth.ItemIndex  := 15;
  cbHeight.ItemIndex := 15;

  SetGUILanguageText;
end;


procedure TfrmImportBitmap.SetGUILanguageText;
begin
  Caption := GLanguageHandler.Text[kImportFromBitmap];

  tsSingleImage.Caption := GLanguageHandler.Text[kSingleImage];
  bSelect.Caption := GLanguageHandler.Text[kSelect];
  lFileName.Caption := GLanguageHandler.Text[kNoImagesSelected];

  tsMultipleImages.Caption := GLanguageHandler.Text[kMultipleImages];
  Label5.Caption := GLanguageHandler.Text[kFirstImage];
  Label8.Caption := GLanguageHandler.Text[kFirstFrame];
  Label10.Caption := GLanguageHandler.Text[kIndexLength];
  lImageLengthExample.Caption := GLanguageHandler.Text[kIndexLengthExample];
  Label7.Caption := GLanguageHandler.Text[kPattern];

  gbSettings.Caption := GLanguageHandler.Text[kImportSettings];
  cbRGBImport.Caption := GLanguageHandler.Text[kRGBImport];
  cbCreateNew.Caption := GLanguageHandler.Text[kCreateNewMatrixClearsAllData];
  Label2.Caption := GLanguageHandler.Text[kFramesToImport];
  Label3.Caption := GLanguageHandler.Text[kFrameWidth];
  Label4.Caption := GLanguageHandler.Text[kFrameHeight];
  bAuto.Caption := GLanguageHandler.Text[kAuto];

  lWidth.Caption := GLanguageHandler.Text[kWidth];
  lHeight.Caption := GLanguageHandler.Text[kHeight];
  lImageWidth.Caption := GLanguageHandler.Text[kNA];
  lImageHeight.Caption := GLanguageHandler.Text[kNA];

  lHelpText.Caption := GLanguageHandler.Text[kForNonRGBImport];

  bOK.Caption := GLanguageHandler.Text[kOK];
  bCancel.Caption := GLanguageHandler.Text[kCancel];
end;


procedure TfrmImportBitmap.sbMISelectFirstImageClick(Sender: TObject);
begin
  if opdMain.Execute then begin
    iMultipleImages.Picture.LoadFromFile(opdMain.FileName);

    cbWidth.Text       := IntToStr(iMultipleImages.Width);
    cbHeight.Text      := IntToStr(iMultipleImages.Height);

    eMIFirstImage.Text := opdMain.FileName;

    SetMultipleImageDetails;

    bOK.Enabled := True;
  end;
end;

procedure TfrmImportBitmap.bOKClick(Sender: TObject);
begin
  ImportMode := ImportModeInvalid;

  case pcImportMethod.ActivePageIndex of
    0 : begin
          if ImageFilename <> '' then begin
            ImportMode  := ImportModeSingleImage;

            FrameCount  := StrToInt(eFrames.Text);
            FrameWidth  := StrToInt(cbWidth.Text);
            FrameHeight := StrToInt(cbHeight.Text);

            RGBImport   := cbRGBImport.Checked;
            CreateNew   := cbCreateNew.Checked;
          end;
        end;
    1 : begin
          if eMIFirstImage.Text <> '' then begin
            ImportMode  := ImportModeMultipleImages;

            FrameCount  := StrToInt(eFrames.Text);
            FrameWidth  := StrToInt(cbWidth.Text);
            FrameHeight := StrToInt(cbHeight.Text);

            FirstFrame  := StrToInt(eMIFirstFrame.Text);
            PadLength   := StrToInt(eMIPadLength.Text);
            Pattern     := eMIPattern.Text;

            RGBImport   := cbRGBImport.Checked;
            CreateNew   := cbCreateNew.Checked;
          end;
        end;
  end;

end;


procedure TfrmImportBitmap.cbRGBImportClick(Sender: TObject);
begin
  lHelpText.Visible := not cbRGBImport.Checked;
end;


procedure TfrmImportBitmap.bCancelClick(Sender: TObject);
begin
  FrameCount := -1;
end;


procedure TfrmImportBitmap.bSelectClick(Sender: TObject);
begin
  if opdMain.Execute then begin
    if FileExists(opdMain.Filename) then begin
      iImport.Picture.LoadFromFile(opdMain.Filename);

      ImageFilename        := opdMain.Filename;

      lFileName.Caption    := opdMain.Filename;

      lImageWidth.Caption  := IntToStr(iImport.Picture.Width);
      lImageHeight.Caption := IntToStr(iImport.Picture.Height);

      bOK.Enabled := True;
    end;
  end;
end;


procedure TfrmImportBitmap.bAutoClick(Sender: TObject);
var
  lFrameCount : integer;

begin
  if (pcImportMethod.ActivePageIndex = 0) then begin

    lFrameCount   := StrToIntDef(eFrames.Text, 1);

    cbWidth.Text  := IntToStr(Round(iImport.Picture.Width / lFrameCount));
    cbHeight.Text := IntToStr(iImport.Picture.Height);
  end;
end;


procedure TfrmImportBitmap.SetMultipleImageDetails;
var
  t : integer;
  lFileName : string;
  lInNumber : boolean;
  lFirstFrame : string;
  lPatternEnd : integer;
  lPatternStart : integer;

begin
  lInNumber := False;
  lFileName := ExtractFileName(eMIFirstImage.Text);

  lPatternStart := -1;
  lPatternEnd   := -1;

  for t := length(lFileName) downto 1 do begin
    if (lFileName[t] >= '0') and (lFileName[t] <= '9') then begin
      if lInNumber then
        lFirstFrame := lFileName[t] + lFirstFrame
      else begin
        lInNumber := True;

        lFirstFrame := lFileName[t];

        lPatternEnd := t;
      end;
    end
    else
      if lInNumber then begin
        lPatternStart := t;

        Break;
      end;
  end;

  if (lPatternStart <> -1) and (lPatternEnd <> -1) then begin
    eMIFirstFrame.Text := lFirstFrame;

    eMIPattern.Text    := ExtractFilePath(eMIFirstImage.Text) + Copy(lFileName, 1, lPatternStart) + '$$' + Copy(lFileName, lPatternEnd + 1);

    eMIPadLength.Text  := IntToStr(lPatternEnd - lPatternStart);
  end;
end;


end.

File: Source/Forms/FormImportBitmap.cpp
// ===================================================================
//
//   (c) Paul Alan Freshney 2012-2024
//   www.freshney.org :: paul@freshney.org :: maximumoctopus.com
//
//   https://github.com/MaximumOctopus/LEDMatrixStudio
//
//   https://maximumoctopus.hashnode.dev/
//
//   C++ Rewrite October 11th 2023
//
// ===================================================================

#include <vcl.h>
#pragma hdrstop

#include "FormImportBitmap.h"
#include "LanguageConstants.h"
#include "LanguageHandler.h"
#include "Utility.h"

extern LanguageHandler *GLanguageHandler;

//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.dfm"
TfrmImportBitmap *frmImportBitmap;
//---------------------------------------------------------------------------

__fastcall TfrmImportBitmap::TfrmImportBitmap(TComponent* Owner)
	: TForm(Owner)
{
	for (int x = 1; x < 65; x++)
	{
		cbWidth->Items->Add(x);
		cbHeight->Items->Add(x);
	}

	cbWidth->ItemIndex  = 15;
	cbHeight->ItemIndex = 15;

	SetGuiLanguageText();
}


void __fastcall TfrmImportBitmap::bOKClick(TObject *Sender)
{
	Import = ImportMode::kInvalid;

	switch (cbImportColourMode->ItemIndex)
	{
		case 0:
			ImportMode = ImportColourMode::kMono;
			break;
		case 1:
			ImportMode = ImportColourMode::kRGB;
			break;
		case 2:
			ImportMode = ImportColourMode::kRGB3bpp;
			break;
	}

	switch (pcImportMethod->ActivePageIndex)
	{
	case 0:
		if (!ImageFilename.empty())
		{
			Import = ImportMode::kSingleImage;

			FrameCount  = eFrames->Text.ToIntDef(-1);
			FrameWidth  = cbWidth->Text.ToIntDef(0);
			FrameHeight = cbHeight->Text.ToIntDef(0);

			CreateNew   = cbCreateNew->Checked;
		}
		break;
	case 1:
		if (eMIFirstImage->Text != L"")
		{
			Import = ImportMode::kMultipleImages;

			FrameCount  = eFrames->Text.ToIntDef(-1);
			FrameWidth  = cbWidth->Text.ToIntDef(0);
			FrameHeight = cbHeight->Text.ToIntDef(0);

			FirstFrame  = eMIFirstFrame->Text.ToIntDef(0);
			PadLength   = eMIPadLength->Text.ToIntDef(0);
			Pattern     = eMIPattern->Text.ToIntDef(0);

			CreateNew   = cbCreateNew->Checked;
		}
		break;
	}
}

void __fastcall TfrmImportBitmap::bCancelClick(TObject *Sender)
{
	FrameCount = -1;
}


void __fastcall TfrmImportBitmap::bSelectClick(TObject *Sender)
{
	if (opdMain->Execute())
	{
		if (FileExists(opdMain->FileName))
		{
			iImport->Picture->LoadFromFile(opdMain->FileName);

			ImageFilename         = opdMain->FileName;

			lFileName->Caption    = opdMain->FileName;

			lImageWidth->Caption  = iImport->Picture->Width;
			lImageHeight->Caption = iImport->Picture->Height;

            bAutoClick(nullptr);

			bOK->Enabled = true;
		}
	}
}


void __fastcall TfrmImportBitmap::bAutoClick(TObject *Sender)
{
	if (pcImportMethod->ActivePageIndex == 0)
	{
		int count = eFrames->Text.ToIntDef(1);

		int width = std::round((double)iImport->Picture->Width / (double)count);

		cbWidth->Text  = width;
		cbHeight->Text = iImport->Picture->Height;
	}
}


void TfrmImportBitmap::SetGuiLanguageText()
{
	Caption = GLanguageHandler->Text[kImportFromBitmap].c_str();

	tsSingleImage->Caption = GLanguageHandler->Text[kSingleImage].c_str();
	bSelect->Caption = GLanguageHandler->Text[kSelect].c_str();
	lFileName->Caption = GLanguageHandler->Text[kNoImagesSelected].c_str();

	tsMultipleImages->Caption = GLanguageHandler->Text[kMultipleImages].c_str();
	Label5->Caption = GLanguageHandler->Text[kFirstImage].c_str();
	Label8->Caption = GLanguageHandler->Text[kFirstFrame].c_str();
	Label10->Caption = GLanguageHandler->Text[kIndexLength].c_str();
	lImageLengthExample->Caption = GLanguageHandler->Text[kIndexLengthExample].c_str();
	Label7->Caption = GLanguageHandler->Text[kPattern].c_str();

	gbSettings->Caption = GLanguageHandler->Text[kImportSettings].c_str();
	cbCreateNew->Caption = GLanguageHandler->Text[kCreateNewMatrixClearsAllData].c_str();
	Label2->Caption = GLanguageHandler->Text[kFramesToImport].c_str();
	Label3->Caption = GLanguageHandler->Text[kFrameWidth].c_str();
	Label4->Caption = GLanguageHandler->Text[kFrameHeight].c_str();
	bAuto->Caption = GLanguageHandler->Text[kAuto].c_str();

	lWidth->Caption = GLanguageHandler->Text[kWidth].c_str();
	lHeight->Caption = GLanguageHandler->Text[kHeight].c_str();
	lImageWidth->Caption = GLanguageHandler->Text[kNA].c_str();
	lImageHeight->Caption = GLanguageHandler->Text[kNA].c_str();

	lHelpText->Caption = GLanguageHandler->Text[kForNonRGBImport].c_str();

	bOK->Caption = GLanguageHandler->Text[kOK].c_str();
	bCancel->Caption = GLanguageHandler->Text[kCancel].c_str();
}


void __fastcall TfrmImportBitmap::sbMISelectFirstImageClick(TObject *Sender)
{
	if (opdMain->Execute())
	{
		iMultipleImages->Picture->LoadFromFile(opdMain->FileName);

		cbWidth->Text       = iMultipleImages->Width;
		cbHeight->Text      = iMultipleImages->Height;

		eMIFirstImage->Text = opdMain->FileName;

		SetMultipleImageDetails();

		bOK->Enabled = true;
	}
}


void TfrmImportBitmap::SetMultipleImageDetails()
{
	bool InNumber = false;
	std::wstring FileName = ExtractFileName(eMIFirstImage->Text).c_str();

	int PatternStart = -1;
	int PatternEnd   = -1;

	for (int t = FileName.length() - 1; t >= 0; t--)
	{
		if (isdigit(FileName[t]))
		{
			if (InNumber)
			{
				FirstFrame = FileName[t] + FirstFrame;
			}
			else
			{
				InNumber = true;

				FirstFrame = FileName[t];

				PatternEnd = t;
			}
		}
		else
		{
			if (InNumber)
			{
				PatternStart = t;

				break;
			}
		}
	}

	if (PatternStart != -1  && PatternEnd != -1)
	{
		eMIFirstFrame->Text = FirstFrame;

		eMIPattern->Text    = ExtractFilePath(eMIFirstImage->Text) + Utility::WS2US(FileName.substr(0, PatternStart) + L"$$" + FileName.substr(0, PatternEnd + 1));

		eMIPadLength->Text  =(PatternEnd - PatternStart);
	}
}


void __fastcall TfrmImportBitmap::cbImportColourModeChange(TObject *Sender)
{
	lHelpText->Visible = cbImportColourMode->ItemIndex > 0;
}
